DetectorDescription/DDCMS/src/DDFilteredView.cc

```diff
@@ -572,7 +572,7 @@ std::string_view DDFilteredView::get<string_view>(const string& key) {
 
 template <>
 double DDFilteredView::get<double>(const string& key) {
-  double result(0.0);
+  double result(std::nan("0"));
```

Comment:
How does a nan help us? A nan can propagate quite far before it causes a crash, and it can be difficult to determine the cause of the crash.

Reply:
Does a `std::optional<double>` make more sense?

Reply:
> Does a `std::optional<double>` make more sense?

hmm... it would require a check on `nullopt` and then throw? 

`nan` is more performant, IMHO.  I know it's an overkill here, but... we do not have to check for errors after each step in a series of calculations if we can be sure that any error will show up in the end result as `nan`. I think, the usage of a numeric value in this case will :-)

@cvuosalo - The advantages of relying on the propagation of `inf` and `nan` are:
• The performance is excellent because no resources are spent on error checking, except for the final result. Current hardware does not take extra time to process `inf` and `nan` inputs.
• The code can be converted to SIMD without any change in the behavior.
• Extra information contained in a `nan` payload will propagate to the end result in most cases. The payload can be used for information about an error. For example, library functions can return an error code in the payload of a `nan` result.

Reply:
How does SIMD work? How would we use it? How would crashes in standard workflows be prevented?

Reply:
Naive question: why this function could not throw if the result is not found?

I partially agree with @ianna's comments, but I've found tracing the origin of NaN's is somewhat painful (even if the FPE option is there).

Reply:
@makortel - it could, the idea of this PR is to shake the tree :-) We have to edit the XMLs

Reply:
Is this PR just for testing? If Matti, with his tremendous expertise, says tracing NaNs is painful, that sounds scary. I don't think we want a NaN in production code.

Reply:
I wouldn’t worry about `nan`. It looks like there is a bigger problem with the lock dependencies, please, see the report. We must get rid of the unit conversions ASAP.

Reply:
@ianna Do you mean the test failure in this PR? It is in step2 of 11624.911, running in single-threaded mode. My own tests of 11624.911 show it runs successfully through step2 in multi-threaded mode, so I'm not sure what problem there is with locks. Wouldn't the first guess for the cause of the test failure be the PR itself?

Reply:
@cvuosalo - Segmentation faults in SSE operations are due to misaligned memory:

08-Dec-2020 18:24:57 CET  Initiating request to open file file:step1.root
08-Dec-2020 18:24:58 CET  Successfully opened file file:step1.root
%MSG-w DeleteModules:  AfterModConstruction 08-Dec-2020 18:26:03 CET pre-events
Following modules are not in any Path or EndPath, nor is their output consumed by any other module, and therefore they are deleted before beginJob transition.
 omtfStage2Raw
 simHcalTTPDigis
%MSG
PersistencyIO    INFO  +++ Set Streamer to dd4hep::OpaqueDataBlock
CompactLoader    INFO  +++ Processing compact file: /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-08-1100/src/Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml with flag (null)
DD4CMS           INFO  +++ Processing the CMS detector description file:///cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-08-1100/src/Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml
Detector         INFO  *********** Created World volume with size: 10100 10100 45000
Detector         INFO  +++ Patching names of anonymous shapes....
DDDefinition     INFO  +++ Finished processing file:///cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2020-12-08-1100/src/Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml


A fatal system signal has occurred: segmentation violation
The following is the call stack containing the origin of the signal.

Tue Dec  8 18:26:38 CET 2020
Thread 2 (Thread 0x2b42cfac0700 (LWP 16508)):
#0  0x00002b42ae4ea1d9 in waitpid () from /lib64/libpthread.so.0
#1  0x00002b42b46b5f97 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#2  0x00002b42b46b69da in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#3  0x00002b42ae0b3af0 in std::execute_native_thread_routine (__p=0x2b42b90cd640) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#4  0x00002b42ae4e2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00002b42ae7f596d in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x2b42b03f3c80 (LWP 16415)):
#0  0x00002b42ae7eac3d in poll () from /lib64/libc.so.6
#1  0x00002b42b46b6397 in full_read.constprop () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#2  0x00002b42b46b6aac in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#3  0x00002b42b46b77d4 in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00002b42ae84b919 in __memmove_ssse3 () from /lib64/libc.so.6
#6  0x00002b42d3007dc4 in CSCGeometryParsFromDD::build(cms::DDCompactView const*, MuonGeometryConstants const&, RecoIdealGeometry&, CSCRecoDigiParameters&) () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/libGeometryCSCGeometryBuilder.so
#7  0x00002b42d3006fe0 in CSCGeometryBuilderFromDDD::build(CSCGeometry&, cms::DDCompactView const*, MuonGeometryConstants const&) () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/libGeometryCSCGeometryBuilder.so
#8  0x00002b42d2d86e54 in CSCGeometryESModule::initCSCGeometry_(MuonGeometryRecord const&, std::shared_ptr<edm::ESProductHost<CSCGeometry, IdealGeometryRecord, CSCRecoGeometryRcd, CSCRecoDigiParametersRcd> >&) () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginCSCGeometryPlugins.so
#9  0x00002b42d2d883a0 in CSCGeometryESModule::produce(MuonGeometryRecord const&) () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginCSCGeometryPlugins.so
#10 0x00002b42d2d95791 in decltype ({parm#1}()) edm::convertException::wrap<edm::eventsetup::Callback<CSCGeometryESModule, std::shared_ptr<CSCGeometry>, MuonGeometryRecord, edm::eventsetup::CallbackSimpleDecorator<MuonGeometryRecord> >::runProducerAsync(std::__exception_ptr::exception_ptr const*, edm::eventsetup::EventSetupRecordImpl const*, edm::EventSetupImpl const*, edm::ServiceToken const&)::{lambda()#1}::operator()() const::{lambda()#1}>(edm::eventsetup::Callback<CSCGeometryESModule, std::shared_ptr<CSCGeometry>, MuonGeometryRecord, edm::eventsetup::CallbackSimpleDecorator<MuonGeometryRecord> >::runProducerAsync(std::__exception_ptr::exception_ptr const*, edm::eventsetup::EventSetupRecordImpl const*, edm::EventSetupImpl const*, edm::ServiceToken const&)::{lambda()#1}::operator()() const::{lambda()#1}) () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginCSCGeometryPlugins.so
#11 0x00002b42d2d95ad0 in edm::eventsetup::Callback<CSCGeometryESModule, std::shared_ptr<CSCGeometry>, MuonGeometryRecord, edm::eventsetup::CallbackSimpleDecorator<MuonGeometryRecord> >::runProducerAsync(std::__exception_ptr::exception_ptr const*, edm::eventsetup::EventSetupRecordImpl const*, edm::EventSetupImpl const*, edm::ServiceToken const&)::{lambda()#1}::operator()() const () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginCSCGeometryPlugins.so
#12 0x00002b42d2d96adf in void edm::SerialTaskQueueChain::actionToRun<edm::eventsetup::Callback<CSCGeometryESModule, std::shared_ptr<CSCGeometry>, MuonGeometryRecord, edm::eventsetup::CallbackSimpleDecorator<MuonGeometryRecord> >::runProducerAsync(std::__exception_ptr::exception_ptr const*, edm::eventsetup::EventSetupRecordImpl const*, edm::EventSetupImpl const*, edm::ServiceToken const&)::{lambda()#1}&>(edm::eventsetup::Callback<CSCGeometryESModule, std::shared_ptr<CSCGeometry>, MuonGeometryRecord, edm::eventsetup::CallbackSimpleDecorator<MuonGeometryRecord> >::runProducerAsync(std::__exception_ptr::exception_ptr const*, edm::eventsetup::EventSetupRecordImpl const*, edm::EventSetupImpl const*, edm::ServiceToken const&)::{lambda()#1}&) () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginCSCGeometryPlugins.so
#13 0x00002b42d2d96b51 in edm::SerialTaskQueue::QueuedTask<edm::SerialTaskQueueChain::push<edm::eventsetup::Callback<CSCGeometryESModule, std::shared_ptr<CSCGeometry>, MuonGeometryRecord, edm::eventsetup::CallbackSimpleDecorator<MuonGeometryRecord> >::runProducerAsync(std::__exception_ptr::exception_ptr const*, edm::eventsetup::EventSetupRecordImpl const*, edm::EventSetupImpl const*, edm::ServiceToken const&)::{lambda()#1}>(edm::eventsetup::Callback<CSCGeometryESModule, std::shared_ptr<CSCGeometry>, MuonGeometryRecord, edm::eventsetup::CallbackSimpleDecorator<MuonGeometryRecord> >::runProducerAsync(std::__exception_ptr::exception_ptr const*, edm::eventsetup::EventSetupRecordImpl const*, edm::EventSetupImpl const*, edm::ServiceToken const&)::{lambda()#1}&&)::{lambda()#1}>::execute() () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/pluginCSCGeometryPlugins.so
#14 0x00002b42ad6d80f5 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::process_bypass_loop (this=this@entry=0x2b42b11e3200, context_guard=..., t=0x2b4309241640, isolation=isolation@entry=0) at ../../src/tbb/scheduler.h:630
#15 0x00002b42ad6d83ab in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x2b42b11e3200, parent=..., child=<optimized out>) at ../../include/tbb/task.h:1003
#16 0x00002b42abfa8255 in tbb::interface7::internal::delegated_function<edm::Worker::esPrefetchAsync(edm::WaitingTask*, edm::EventSetupImpl const&, edm::Transition, edm::ServiceToken const&)::{lambda()#1}::operator()()::{lambda()#1} const, void>::operator()() const () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#17 0x00002b42ad6d2c42 in tbb::interface7::internal::task_arena_base::internal_execute (this=0x2b42ac09ef40 <edm::esTaskArena()::s_arena>, d=...) at ../../src/tbb/arena.cpp:1105
#18 0x00002b42abfa7a95 in edm::FunctorTask<edm::Worker::esPrefetchAsync(edm::WaitingTask*, edm::EventSetupImpl const&, edm::Transition, edm::ServiceToken const&)::{lambda()#1}>::execute() () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#19 0x00002b42ad6d80f5 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::process_bypass_loop (this=this@entry=0x2b42b11e3200, context_guard=..., t=0x2b4309242640, isolation=isolation@entry=0) at ../../src/tbb/scheduler.h:630
#20 0x00002b42ad6d83ab in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x2b42b11e3200, parent=..., child=<optimized out>) at ../../include/tbb/task.h:1003
#21 0x00002b42abe996b7 in edm::EventProcessor::beginRun(edm::Hash<2> const&, unsigned int, bool&, bool&) () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#22 0x00002b42abe9a027 in edm::EventProcessor::runToCompletion() () from /cvmfs/cms-ib.cern.ch/nweek-02658/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2020-12-06-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#23 0x000000000040fb86 in tbb::interface7::internal::delegated_function<main::{lambda()#1}::operator()() const::{lambda()#1} const, void>::operator()() const ()
#24 0x00002b42ad6d2c42 in tbb::interface7::internal::task_arena_base::internal_execute (this=0x7ffd0a6cbcd0, d=...) at ../../src/tbb/arena.cpp:1105
#25 0x0000000000410aa3 in main::{lambda()#1}::operator()() const ()
#26 0x000000000040f6dc in main ()

Reply:
> Naive question: why this function could not throw if the result is not found?
> 
done